File: src/main/java/cspom/compiler/patterns/AllDiff.java
package cspom.compiler.patterns;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import cspom.compiler.ConstraintSelector;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.GeneralConstraint;
import cspom.CSPOM;
import cspom.variable.CSPOMVariable;

public final class AllDiff implements ConstraintCompiler {

    private static final ConstraintSelector DIFF_CONSTRAINT = new ConstraintSelector() {
        @Override
        public boolean is(final CSPOMConstraint constraint) {
            return constraint instanceof GeneralConstraint
                    && "ne".equals(constraint.description())
                    || "gt".equals(constraint.description())
                    || "lt".equals(constraint.description())
                    || "allDifferent".equals(constraint.description());
        }
    };

    private static final ConstraintSelector ALLDIFF_CONSTRAINT = new ConstraintSelector() {
        @Override
        public boolean is(final CSPOMConstraint constraint) {
            return constraint instanceof GeneralConstraint
                    && "ne".equals(constraint.description())
                    || "allDifferent".equals(constraint.description());
        }
    };

    private static final int ITER = 750;
    private static final Random RAND = new Random(0);

    private final CSPOM problem;
    private final CliqueDetector cliqueDetector;

    public AllDiff(final CSPOM problem) {
        this.problem = problem;
        this.cliqueDetector = new CliqueDetector();
    }

    /**
     * If constraint is part of a larger clique of inequalities, replace it by a
     * larger all-diff constraint.
     * 
     * @param constraint
     */
    public void alldiff(final CSPOMConstraint constraint) {
        if (!DIFF_CONSTRAINT.is(constraint)) {
            return;
        }
        // System.out.print(constraint);

        final Set<CSPOMVariable<?>> pool = new HashSet<CSPOMVariable<?>>();
        final Set<CSPOMVariable<?>> clique = new HashSet<CSPOMVariable<?>>(
                JavaCollections.asJavaList(constraint.scope()));
        populate(pool, clique);
        expand(clique, pool);

        if (clique.size() > constraint.scope().size()) {
            // System.out.print(" -> " + clique.size() + "-clique");
            newAllDiff(clique);
            // System.out.print(", " + (rem.size() - 1) +
            // " constraints removed, "
            // + problem.getConstraints().size() + " remaining");
        }
        // System.out.println();

    }

    private void populate(final Set<CSPOMVariable<?>> pool,
            final Collection<CSPOMVariable<?>> base) {
        final Iterator<CSPOMVariable> baseItr = base.iterator();
        if (!baseItr.hasNext()) {
            return;
        }
        final CSPOMVariable first = baseItr.next();
        for (CSPOMConstraint c : first.getConstraints()) {
            if (DIFF_CONSTRAINT.is(c)) {
                for (CSPOMVariable n : c.getScope()) {
                    if (n != first) {
                        pool.add(n);
                    }
                }
            }
        }
        pool.removeAll(base);
        while (baseItr.hasNext()) {
            final CSPOMVariable variable = baseItr.next();
            for (Iterator<CSPOMVariable> poolItr = pool.iterator(); poolItr
                    .hasNext();) {
                final CSPOMVariable poolVar = poolItr.next();
                if (!cliqueDetector.edge(variable, poolVar, DIFF_CONSTRAINT)) {
                    poolItr.remove();
                }
            }
        }
    }

    private static final int TABU_SIZE = 15;

    private static CSPOMVariable<?> pick(final Set<CSPOMVariable<?>> pool,
            final Map<CSPOMVariable<?>, Integer> tabu, final int iteration) {
        int tie = 1;
        CSPOMVariable returned = null;
        for (CSPOMVariable v : pool) {
            final Integer tabuLength = tabu.get(v);
            if ((tabuLength == null || tabuLength > iteration)
                    && RAND.nextFloat() * tie++ < 1) {
                returned = v;
            }
        }

        if (returned == null) {
            return null;
        }
        tabu.put(returned, iteration - TABU_SIZE);
        return returned;
    }

    private void expand(final Set<CSPOMVariable<?>> clique,
            final Set<CSPOMVariable<?>> pool) {
        final Set<CSPOMVariable> largest = new HashSet<CSPOMVariable>(clique);
        final Set<CSPOMVariable> base = new HashSet<CSPOMVariable>(clique);
        final Map<CSPOMVariable, Integer> tabu = new HashMap<CSPOMVariable, Integer>();

        for (int i = ITER; --i >= 0;) {
            final CSPOMVariable newVar = pick(pool, tabu, i);
            if (newVar == null) {
                if (clique.size() - base.size() <= 0) {
                    break;
                }
                int rand = RAND.nextInt(clique.size() - base.size());
                CSPOMVariable toRemove = null;
                for (CSPOMVariable v : clique) {
                    if (!base.contains(v) && --rand < 0) {
                        toRemove = v;
                        break;
                    }
                }
                clique.remove(toRemove);
                pool.clear();
                populate(pool, clique);
            } else {
                clique.add(newVar);
                if (clique.size() > largest.size()) {
                    largest.clear();
                    largest.addAll(clique);
                }
                pool.remove(newVar);
                for (Iterator<CSPOMVariable> itr = pool.iterator(); itr
                        .hasNext();) {
                    if (!cliqueDetector.edge(itr.next(), newVar,
                            DIFF_CONSTRAINT)) {
                        itr.remove();
                    }
                }
            }
        }

        clique.clear();
        clique.addAll(largest);
    }

    /**
     * Adds a new all-diff constraint of the specified scope. Any newly subsumed
     * neq/all-diff constraints are removed.
     * 
     * @param scope
     */
    public void newAllDiff(final Collection<CSPOMVariable<?>> scope) {
        final CSPOMConstraint allDiff = new GeneralConstraint("allDifferent",
                null, scope.toArray(new CSPOMVariable[scope.size()]));
        if (problem.getConstraints().contains(allDiff)) {
            return;
        }
        problem.addConstraint(allDiff);

        /*
         * Remove newly subsumed neq/alldiff constraints.
         */
        final Set<CSPOMConstraint> subsumed = new HashSet<CSPOMConstraint>();
        for (CSPOMVariable v : scope) {
            for (CSPOMConstraint c : v.getConstraints()) {
                if (c != allDiff && ALLDIFF_CONSTRAINT.is(c)) {
                    subsumed.add(c);
                }
            }
        }

        for (Iterator<CSPOMConstraint> itr = subsumed.iterator(); itr.hasNext();) {
            final CSPOMConstraint candidate = itr.next();
            if (!CliqueDetector.subsumes(allDiff, candidate)) {
                itr.remove();
            }
        }
        for (CSPOMConstraint c : subsumed) {
            problem.removeConstraint(c);
        }
    }

    /**
     * If the given constraint is an all-different or neq constraint, remove it
     * if it is subsumed by another difference constraint.
     * 
     * @param constraint
     */
    public boolean dropSubsumedDiff(final CSPOMConstraint constraint) {
        if (ALLDIFF_CONSTRAINT.is(constraint)
                && CliqueDetector.haveSubsumingConstraint(constraint,
                        DIFF_CONSTRAINT)) {
            problem.removeConstraint(constraint);
            return true;
        }
        return false;
    }

    @Override
    public void compile(final CSPOMConstraint constraint) {
        if (!dropSubsumedDiff(constraint)) {
            alldiff(constraint);
        }

    }

}

File: src/main/scala/cspom/compiler/patterns/AllDiff.scala
package cspom.compiler.patterns;

import scala.collection.JavaConversions
import cspom.CSPOM
import cspom.compiler.ConstraintSelector
import cspom.constraint.CSPOMConstraint
import cspom.constraint.GeneralConstraint
import cspom.variable.CSPOMVariable
import scala.util.Random
import scala.util.control.Breaks

final class AllDiff(val problem: CSPOM) extends ConstraintCompiler {

  val neighbors =
    problem.variables map (v =>
      v -> (v.constraints.iterator.filter(DIFF_CONSTRAINT).foldLeft(Set[CSPOMVariable]())(
        (acc, c) => acc ++ c.scope) - v)) toMap

  def DIFF_CONSTRAINT(constraint: CSPOMConstraint) =
    constraint.isInstanceOf[GeneralConstraint] &&
      Set("ne", "gt", "lt", "allDifferent").contains(constraint.description)

  def ALLDIFF_CONSTRAINT(constraint: CSPOMConstraint) =
    constraint.isInstanceOf[GeneralConstraint] &&
      "ne" == constraint.description ||
      "allDifferent" == constraint.description

  val ITER = 750;

  //val cliqueDetector = new CliqueDetector;

  /**
   * If constraint is part of a larger clique of inequalities, replace it by a
   * larger all-diff constraint.
   *
   * @param constraint
   */
  def alldiff(constraint: CSPOMConstraint) {
    if (!DIFF_CONSTRAINT(constraint)) {
      return ;
    }
    // System.out.print(constraint);

    //val clique: Set[CSPOMVariable] = Set.empty ++ constraint.scope

    // val pool = populate(constraint.scope);
    //print(constraint + " : ")
    val clique = expand(constraint.scope.toSet);
    //println(clique.size)
    if (clique.size > constraint.scope.size) {
      problem.removeConstraint(constraint)
      newAllDiff(clique);
    }

  }

  /**
   * The pool contains all variables that can expand the base clique
   */
  private def populate(base: Set[CSPOMVariable]): Set[CSPOMVariable] =
    base.iterator.map(neighbors(_)).reduceLeft((acc, vs) => acc & vs)

  private def expand(base: Set[CSPOMVariable]) = {

    var largest = base
    var clique = base
    var pool = populate(base)

    //final Set<CSPOMVariable> base = new HashSet<CSPOMVariable>(clique);

    var tabu: Map[CSPOMVariable, Int] = Map.empty
    val mybreaks = new Breaks
    import mybreaks.{ break, breakable }

    breakable {
      for (i <- (1 to ITER)) {
        val (newVar, newTabu) = AllDiff.pick(pool, tabu, i);
        tabu = newTabu
        newVar match {
          case None => {
            /* Could not expand the clique, removing a variable (not from the base) */
            AllDiff.pick((clique -- base).iterator) match {
              case None => break
              case Some(variable) => clique -= variable
            }
            pool = populate(clique)
          }
          case Some(variable) => {
            clique += variable
            if (clique.size > largest.size) {
              largest = clique
            }
            pool -= variable;
            pool &= neighbors(variable)

          }
        }
        //println(System.currentTimeMillis + " : " + clique.size)
      }
    }

    largest
  }

  /**
   * Adds a new all-diff constraint of the specified scope. Any newly subsumed
   * neq/all-diff constraints are removed.
   *
   * @param scope
   */
  private def newAllDiff(scope: Set[CSPOMVariable]) {
    val allDiff = new GeneralConstraint(
      description = "allDifferent",
      scope = scope.toSeq);

    if (problem.constraints.contains(allDiff)) {
      return ;
    }
    problem.addConstraint(allDiff);

    //    var removed = 0
    //    /* Remove newly subsumed neq/alldiff constraints. */
    //    for (
    //      v <- scope; c <- v.constraints if (c != allDiff && ALLDIFF_CONSTRAINT(c) && CliqueDetector.subsumes(allDiff, c))
    //    ) {
    //      removed += 1
    //      problem.removeConstraint(c);
    //    }
    //    println(System.currentTimeMillis + " : removed " + removed + " constraints, " + problem.constraints.size + " left")
  }

  /**
   * If the given constraint is an all-different or neq constraint, remove it
   * if it is subsumed by another difference constraint.
   *
   * @param constraint
   */
  def dropSubsumedDiff(constraint: CSPOMConstraint) = {
    if (ALLDIFF_CONSTRAINT(constraint) &&
      CliqueDetector.haveSubsumingConstraint(constraint, DIFF_CONSTRAINT)) {

      problem.removeConstraint(constraint);
      //println("subsumed ! " + problem.constraints.size + " remaining")
      true;
    } else {
      false;
    }
  }

  override def compile(constraint: CSPOMConstraint) {
    if (!dropSubsumedDiff(constraint)) {
      alldiff(constraint);
    }

  }

}

object AllDiff {
  val RAND = new Random(0);

  val TABU_SIZE = 15;

  private def pick(pool: Set[CSPOMVariable], tabu: Map[CSPOMVariable, Int], iteration: Int): (Option[CSPOMVariable], Map[CSPOMVariable, Int]) = {

    pick(pool.iterator.filter(v =>
      tabu.get(v) match {
        case None => true
        case Some(i) => i < iteration
      })) match {
      case None =>
        (None, tabu)
      case Some(v) =>
        (Some(v), tabu + (v -> (iteration + TABU_SIZE)))
    }

  }

  private def pick[T](it: Iterator[T]): Option[T] = {
    var tie = 1
    var returned: Option[T] = None
    for (i <- it) {
      if (RAND.nextDouble() * tie < 1) returned = Some(i)
      tie += 1
    }
    returned
  }
}
